File: wrkr-lua/lua-stubs/wrkr/shared.lua
---@meta

---@class wrkr.shared
local M = {}

---Get a value from the shared store.
---@param key string
---@return any|nil
function M.get(key)
	return nil
end

---Set a value in the shared store (JSON-encoded).
---@param key string
---@param value any
function M.set(key, value) end

---Delete a key (and any counter with the same name).
---@param key string
function M.delete(key) end

---Increment a shared counter and return its new value.
---@param key string
---@param delta? integer
---@return integer
function M.incr(key, delta)
	return 0
end

---Get a shared counter value (0 if missing).
---@param key string
---@return integer
function M.counter(key)
	return 0
end

---Wait until a key exists, then return its value.
---@async
---@param key string
---@return any
function M.wait(key)
	return nil
end

---Wait on a named barrier.
---@async
---@param name string
---@param parties integer
function M.barrier(name, parties)
	return nil
end

return M

File: wrkr-lua/src/modules/shared.rs
use std::sync::Arc;

use mlua::{Lua, Value};

mod opts;
mod result;

use opts::SharedSetLuaArgs;
use result::shared_value_to_lua;

pub(super) fn register_runtime(
    lua: &Lua,
    run_ctx: Arc<wrkr_core::RunScenariosContext>,
) -> crate::Result<()> {
    let shared = run_ctx.shared.clone();
    let loader = {
        let shared = shared.clone();
        lua.create_function(move |lua, ()| {
            let t = lua.create_table()?;

            let get = {
                let shared = shared.clone();
                lua.create_function(move |lua, key: String| {
                    shared_value_to_lua(lua, shared.get(&key))
                })?
            };

            let set = {
                let shared = shared.clone();
                lua.create_function(move |lua, (key, value): (String, Value)| {
                    let args = SharedSetLuaArgs::parse(lua, key, value)?;
                    shared.set(&args.key, args.value);
                    Ok(())
                })?
            };

            let delete = {
                let shared = shared.clone();
                lua.create_function(move |_lua, key: String| {
                    shared.delete(&key);
                    Ok(())
                })?
            };

            let incr = {
                let shared = shared.clone();
                lua.create_function(move |_lua, (key, delta): (String, Option<i64>)| {
                    let delta = delta.unwrap_or(1);
                    Ok(shared.incr(&key, delta))
                })?
            };

            let counter = {
                let shared = shared.clone();
                lua.create_function(move |_lua, key: String| Ok(shared.get_counter(&key)))?
            };

            let wait = {
                let shared = shared.clone();
                lua.create_async_function(move |lua, key: String| {
                    let shared = shared.clone();
                    async move {
                        shared.wait_for_key(&key).await;
                        shared_value_to_lua(&lua, shared.get(&key))
                    }
                })?
            };

            let barrier = {
                let shared = shared.clone();
                lua.create_async_function(move |_lua, (name, parties): (String, u64)| {
                    let shared = shared.clone();
                    async move {
                        let parties = parties.min(usize::MAX as u64) as usize;
                        shared
                            .barrier_wait(&name, parties)
                            .await
                            .map_err(mlua::Error::external)?;
                        Ok(())
                    }
                })?
            };

            t.set("get", get)?;
            t.set("set", set)?;
            t.set("delete", delete)?;
            t.set("incr", incr)?;
            t.set("counter", counter)?;
            t.set("wait", wait)?;
            t.set("barrier", barrier)?;

            Ok::<_, mlua::Error>(t)
        })?
    };

    super::preload_set(lua, "wrkr/shared", loader)
}
